### PR TITLE
Paste csv as table feature created for plugin

### PR DESCRIPTION
**Pull request by coderrsid**
Passed tests: 

image

Working Functionality: 
https://user-images.githubusercontent.com/35633575/123890972-cbe33c00-d975-11eb-8a6a-fe238f1c31a9.mov

### CODE REVIEW

File: src/utils/csvAsTable/csvAsTable.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+export const convertCsvToMarkdown = (csv) => {
+	let mdTable: string;
+	mdTable = csvToMarkdown(csv.toString(), ',', true);
+	return mdTable;
+}
+
+export async function csvAsTable (csv: string) {
+	let mdTable: string, parsedCsv: string;
+
+	parsedCsv =  parseCsv(csv);
```
Comment (roman-r-m) [inline src/utils/csvAsTable/csvAsTable.ts]: what does this function do?

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts]: This functions contains the definition to parse csv and return the results if there was no error parsing csv string into array. If there are errors, parsedCsv does not get assigned any value and the code next to that checks if parsedCsv doesn't contain a value, throw an error as the csv string copied from clipboard must not be a proper csv string.

Comment (roman-r-m) [reply src/utils/csvAsTable/csvAsTable.ts]: sure but what do you do with the return value assuming parsing succeeds?

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts]: Oh, actually roman, the code in the csvToMarkdown.ts contains the logic to convert csv string directly to markdown. So, csv parsing is done here so as to check whether the csv we got from clipboard is in proper format or no. So it doesn't make out a table from every type of text from clipboard

Comment (roman-r-m) [reply src/utils/csvAsTable/csvAsTable.ts]: so you're parsing CSV twice, first to validate and then to convert? isn't it a bit excessive?

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts]: No, the clipboard string is called with 
convertCsvToMarkdown() (there is no parsing in this fn), only when it is successfully parsed by parseCsv().

Comment (roman-r-m) [reply src/utils/csvAsTable/csvAsTable.ts]: essentially you are parsing csv to table in `csvToMarkdown`

File: src/utils/csvAsTable/csvToMarkdown.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+/**
+ * Convert CSV to Markdown Table
+ *
+ * @param {string} csvContent - The string content of the CSV
+ * @param {string} delimiter - The character(s) to use as the CSV column delimiter
+ * @param {boolean} hasHeader - Whether to use the first row of Data as headers
+ * @returns {string}
+ */
+ export function csvToMarkdown(csvContent: string, delimiter: string = "\t", hasHeader: boolean = false): string {
```
Comment (laurent22) [inline src/utils/csvAsTable/csvToMarkdown.ts]: This file has the same problem as before - https://github.com/joplin/plugin-paste-special/pull/1/files#r654899142

We had a discussion and:

- You agreed to use PaPaParse - you didn't really use it. I can't say you used it if you're only doing validation with it, and ignore the parsed result.

- You said "i'll use that as a package instead of code." - you didn't do it. You've copied and pasted the code again instead of importing the npm package.

So the result is that you're using two CSV parsing libraries - one that you don't really use, the other that you've copied and pasted instead of importing as an npm package. And the lib you use we've also established that you shouldn't use it because it's nowhere near as good as the PaPaParse one.

For me all this is not really acceptable anymore. I can understand you don't know everything and can make mistakes but we keep repeating the same thing over and over again. Every time you say "yes", and every time you either don't get any of it, or you don't care and keep doing the same thing.

Comment (coderrsid) [reply src/utils/csvAsTable/csvToMarkdown.ts]: >So the result is that you're using two CSV parsing libraries - one that you don't really use, the other that you've copied and pasted instead of importing as an npm package. And the lib you use we've also established that you shouldn't use it because it's nowhere near as good as the PaPaParse one.

Okay, now i understand. It's more clear that the flow should be: 
        CSV --> Papa Parse --> Convert those results to markdown table
        
> You said "i'll use that as a package instead of code." - you didn't do it. You've copied and pasted the code again instead of importing the npm package.

It was not yielding proper results, as it contained only one single code file and tests. But anyways, if i use that package it would parse a csv string first then convert, so should i write a mixed code instead of using that package?

> For me all this is not really acceptable anymore. I can understand you don't know everything and can make mistakes but we keep repeating the same thing over and over again. Every time you say "yes", and every time you either don't get any of it, or you don't care and keep doing the same thing.

Laurent, i know i am making mistakes. But, i am literally not doing this out of spite. I do care about what the mentors guide me. I was not able to understand properly, i guess.

Comment (tessus) [reply src/utils/csvAsTable/csvToMarkdown.ts]: > I was not able to understand properly, i guess.

This is the problem. There's no guessing. If you don't know somthing, you should ask.

We've explained how to approach this several times. Every ttime you say "yes", but do the same thing again or make it worse. The only explanation at this point is that you don't understand the workflow nor the challenges in that workflow.

You don't need your regexes, splitting functions and what else you were doing. Did you even read through the documentation of PapaParse? I doubt it, oterwise you would have come to the conclusion that you don't need "your" parsing code (which is not even yours since you copied it).

So. Look at the package PapaParse. What does it return? An integer? A string? An object?

What can you do with what is returned?

Comment (coderrsid) [reply src/utils/csvAsTable/csvToMarkdown.ts]: Yes, Helmut i agreed earlier too that i copied this code. But using it as a library was not yielding proper results. Instead importing made sense as it just contained this single file in package. I'm gonna mention thanks to that package in the docs absolutely.

Yes, now i understand that the object of array i get from papa-parse, i should use that instead of just validation. I had the idea of using it's results for validation but now i understand i should use those arrays and convert that to markdown.

**Comment by coderrsid** [comment]
Hey everyone, i apologise for not understanding the workflow of plugin you suggested, but in the latest commits you'll see now i'm using papa parse results to create markdown table and now i'm not using a dependency code and written custom code to convert those parsed results to markdown table. 

Also, tried to break my code to know vulnerabilities, added test for most boundary cases. Still consistently testing for more cases. Please do review.

For the format of markdown table in the editor, i am looking forward to use https://github.com/darkriszty/MarkdownTablePrettify-VSCodeExt#readme

File: src/utils/csvAsTable/csvToMarkdown.ts (hunk 2 above)
Comment (tessus) [reply src/utils/csvAsTable/csvToMarkdown.ts]: > Yes, Helmut i agreed earlier too that i copied this code. But using it as a library was not yielding proper results. Instead importing made sense as it just contained this single file in package. I'm gonna mention thanks to that package in the docs absolutely.

Absolutely not. If you still think you need anyhing else but PapaParse, I give up.

> i should use that instead of just validation

Yes, this is what we have been saying all along.

**Comment by tessus** [comment]
Maybe we do have a language barrier problem, I don't know. But this should not be an issue. In that case, instead of saying "yes, I understand" you should tell us that you don't.

All that stuff that we are discussing right now should have been clarified in the bonding period. I also asked for a more detailed timeline a few weeks ago. What happened to that one? (Actually I asked twice.)
Do you understand that you are basically 5 weeks behind? What happened in those 5 weeks?

You have to be way more careful in what you are saying and what you are delivering. If you were part of my team, I'd have already fired you. I don't have an issue with people not being an expert in everything, but if someone says "yes" I should be able to expect a result. As mentioned before, in such a case don't just say "yes", but "no, I don't understand - let's figure it out".
You've been studying CS, so this can't be your first project. 

You have to ask yourself, how would an employer react to such a behavior? In North America, it would not be good. That much I can tell you.

Other than working on a project, GSoC also is a preview of what's to come in real life - out of university. As mentioned before, it's not a problem that someone does not know everything. The fundamentals should be clear though.

So, from now on, never, ever say "yes" again, unless you understand the issue an all of its consequences.

**Comment by roman-r-m** [comment]
> I don't have an issue with people not being an expert in everything, but if someone says "yes" I should be able to expect a result. As mentioned before, in such a case don't just say "yes", but "no, I don't understand - let's figure it out".

If I may add to this -- and this is just my perspective -- it's even ok to say "yes" only to come back later and say "in fact I thought I understood but i's still not clear, can we go through this again?".

File: src/utils/csvAsTable/csvAsTable.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		},
+		error: (err) => {
+			console.info("PapaParse err: ", err[0]?.message);
+		}
+	});
+	
+	return parsedCsv;
+}
+
+const csvToMarkdown = (csvContent): string => {
```
Comment (roman-r-m) [inline src/utils/csvAsTable/csvAsTable.ts]: please don't forget to add type annotations

**Comment by coderrsid** [comment]
> Maybe we do have a language barrier problem, I don't know. But this should not be an issue. In that case, instead of saying "yes, I understand" you should tell us that you don't.
> 
> All that stuff that we are discussing right now should have been clarified in the bonding period. I also asked for a more detailed timeline a few weeks ago. What happened to that one? (Actually I asked twice.)
> Do you understand that you are basically 5 weeks behind? What happened in those 5 weeks?

Agreed, Helmut. I could clearly see that language barrier problem from the very first problem, i faced. Although, thanks to roman. He could understand me a bit more clearly on code bits.
Actually, Helmut. At end, laurent decided for just doing csv as table for this timeline and if time left, we could move ahead. Although, i didn't understand what you wanted to change with timeline?

> You have to be way more careful in what you are saying and what you are delivering. If you were part of my team, I'd have already fired you. I don't have an issue with people not being an expert in everything, but if someone says "yes" I should be able to expect a result. As mentioned before, in such a case don't just say "yes", but "no, I don't understand - let's figure it out".
> You've been studying CS, so this can't be your first project.

Sorry for that. I assure you, i'll do better from this.

> You have to ask yourself, how would an employer react to such a behavior? In North America, it would not be good. That much I can tell you.
> 
> Other than working on a project, GSoC also is a preview of what's to come in real life - out of university. As mentioned before, it's not a problem that someone does not know everything. The fundamentals should be clear though.
> 
> So, from now on, never, ever say "yes" again, unless you understand the issue an all of its consequences.

Absolutely.

File: src/utils/csvAsTable/csvAsTable.ts (hunk 3 above)
Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts]: Done. Thanks roman

**Comment by coderrsid** [comment]
Please do review as soon as possible @laurent22. Thank you so much

File: src/utils/csvAsTable/csvAsTable.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	return parsedCsv;
+}
+
+const csvToMarkdown = (parsedCsv: any[]): string => {
+	// remove pipes and backslashes
+	parsedCsv.forEach((rows, i) => {
+		rows.forEach((e,j) => {
+			parsedCsv[i][j] = e.replace(/[|\\]/g,'');
+		});
+	})
```
Comment (tessus) [inline src/utils/csvAsTable/csvAsTable.ts:25]: Do you have to remove them? Shouldn't they be escaped anyway in the object the papaparse returns?

There are options to specify an escape character and characters that trigger escaping.

File: src/utils/csvAsTable/csvAsTable.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import { parse } from 'papaparse';
+
+const parseCsv = async (csv: string) => {
+	let parsedCsv: [] = undefined;
+	await parse(csv, {
```
Comment (tessus) [inline src/utils/csvAsTable/csvAsTable.ts:5]: have you looked at all options that are available for papaparse?

**Review by tessus — COMMENTED** [review]
The code is a lot better now.

I added a few comments inline.

**Comment by tessus** [comment]
> Although, i didn't understand what you wanted to change with timeline?

I wanted to have a detailed timeline as other students provided. When they work on what, when they deliver, ....
Something like this https://github.com/joplin/plugin-conflict-resolution#timeline

File: src/utils/csvAsTable/csvAsTable.ts (hunk 4 above)
Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:25]: Yes, i tried `escapeChar` and `quoteChar` with obvious options and was not able to get the same results i'm getting here by replacing them using regex.

File: src/utils/csvAsTable/csvAsTable.ts (hunk 5 above)
Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:5]: Yes, i have seen the config(https://www.papaparse.com/docs#config) object and it's different properties. Right now, the only one seemed to be added was `skipEmptyLines`. Should i?

File: src/utils/csvAsTable/csvAsTable.ts (hunk 4 above)
Comment (tessus) [reply src/utils/csvAsTable/csvAsTable.ts:25]: in that case you have to figure out why that is. You might want to open a ticket with papaparse.

IMO using the regex for escaping is only a workaround and should be noted in the code with a comment.

**Comment by coderrsid** [comment]
> > Although, i didn't understand what you wanted to change with timeline?
> 
> I wanted to have a detailed timeline as other students provided. When they work on what, when they deliver, ....
> Something like this https://github.com/joplin/plugin-conflict-resolution#timeline

Sure, i'll create that and add here in readme

File: src/utils/csvAsTable/csvAsTable.ts (hunk 5 above)
Comment (tessus) [reply src/utils/csvAsTable/csvAsTable.ts:5]: I usually like to specify options explicitly, others don't. But you have to think about the pros and cons for options and come up with a plan which you want to use and why.

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:5]: Yes absolutely, i agree. But skipping empty lines, not gonna make sense in every case.

Comment (tessus) [reply src/utils/csvAsTable/csvAsTable.ts:5]: Ok, just to clarify, because I am not 100% sure what you mean.

Does skipping lines make sense or not? Does it make sense sometimes? Currently your sentence means that it does not always make sense. However, if it does not always make sense it's probably better not to use it, unless you have good reasons for it.

But your sentence does not describe any of it, so what is it?

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:5]: User could've left an empty line for a million possible reasons which he/she needs it there in file itself, so skipping wouldn't always makes sense

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:5]: TEST CASE:
`
A, b, c
\Cod, e|f, f
de|\f, \g
`

**Case 1: without quoteChar and escapeChar**

Parsed Result Array:  
[ ["A", " b", " c"], ["\Cod", " e|f", " f"], ["de|\f", " \g"] ]

**Case 2: using config: ` quoteChar: '\\' || '|'`**

Parsed Result Array: 
[ ["A", " b", " c"], ["Cod, e|f, f↵de|\f, \g"] ]

**Case 3: Current Case**

Parsed Result Array: 
[ ["A", " b", " c"], ["Cod", " ef", " f"], ["def", " g"] ]

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:5]: You can see the different results here, Helmut.

File: src/utils/csvAsTable/csvAsTable.ts (hunk 4 above)
Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:25]: TEST CASE:
`
A, b, c
\Cod, e|f, f
de|\f, \g
`

**Case 1: without quoteChar and escapeChar**

Parsed Result Array:  
`[ ["A", " b", " c"], ["\Cod", " e|f", " f"], ["de|\f", " \g"] ]`

**Case 2: using config: ` quoteChar: '\\' || '|'`**

Parsed Result Array: 
`[ ["A", " b", " c"], ["Cod, e|f, f↵de|\f, \g"] ]`

**Case 3: Current Case**

Parsed Result Array: 
`[ ["A", " b", " c"], ["Cod", " ef", " f"], ["def", " g"] ]`

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:25]: You can see observe all different results here, Helmut

File: src/utils/csvAsTable/csvAsTable.ts (hunk 5 above)
Comment (tessus) [reply src/utils/csvAsTable/csvAsTable.ts:5]: > User could've left an empty line for a million possible reasons which he/she needs it there in file itself, so skipping wouldn't always makes sense

Once again, "not always" is a useless description. Either you want to use the option, or you don't. There's no not always. Either way, you have to justify it.

Also, I think you mix up quoteChar and escapeChar.

Comment (tessus) [reply src/utils/csvAsTable/csvAsTable.ts:5]: Another thing. The value `'\\' || '|'` is not valid for quoteChar. You can use either or, but not both.

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:5]: I would not like to use that option, IMO it should skip empty lines. No harm in that. If user wanna delete those, he/she will be able to easily edit it through table.

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:5]: Although, i tested with either of them and the results are same.

Comment (tessus) [reply src/utils/csvAsTable/csvAsTable.ts:5]: > I would not like to use that option, IMO it should skip empty lines.

This sentence is a contradiction in itself. If you are not using the option, lines won't be skipped.

Do you know what this option really does?

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:5]: In here https://www.papaparse.com/docs#config , by default, skipEmptyLines is false. So, i would like to set it to true. That makes more sense?

Comment (coderrsid) [reply src/utils/csvAsTable/csvAsTable.ts:5]: In here https://www.papaparse.com/docs#config , by default, skipEmptyLines is false. So, i would set it to keep it false. As user should have the liberty to give empty rows in their table. That makes more sense?

**Comment by coderrsid** [comment]
Hey @tessus,
I have added the timeline. Please do tell me if it needs changes.

File: src/utils/csvAsTable/csvAsTable.ts (hunk 5 above)
Comment (tessus) [reply src/utils/csvAsTable/csvAsTable.ts:5]: You don't have any test cases for neither skipEmptyLines false nor skipEmptyLines true.

**Comment by tessus** [comment]
I think the timeline should be fine.

File: src/utils/csvAsTable/csvAsTable.ts (hunk 4 above)
Comment (tessus) [reply src/utils/csvAsTable/csvAsTable.ts:25]: I think you need more test cases. 

e.g. what happens:

```
bla,this is very "very" strange,another bla
"very strange "hello" bla", c2, c3
c1,c2,another test \"now\" what
c1,c2,"another test \"now\" what"
```

**Comment by coderrsid** [comment]
Also, do you approve the code? We can merge this as soon as possible and i'll add if more fixes which are there to come as every other project. I also have to create another PR for import CSV after this to stay on timeline

**Comment by tessus** [comment]
I would like to get @laurent22's input first, but I think as a first version it might be ok. I'm still not too happy with the [manual escaping](https://github.com/joplin/plugin-paste-special/pull/2/files#diff-dbaded28d0aac63950577de3ba365815d3f01c538ac3746e364fce47b643dcb4R23). Did you open an issue with PapaParse as I have suggested? (Although I don't believe there's a bug in the code, but that you don't fully understand the quoting and escaping Char options. But this is also something that you can probably find out with a ticket in the papaparse repo.)

**Comment by tessus** [comment]
Please do not force-push. This makes reviewing and following the commits harder as necessary.

**Comment by coderrsid** [comment]
> I would like to get @laurent22's input first, but I think as a first version it might be ok. I'm still not too happy with the [manual escaping](https://github.com/joplin/plugin-paste-special/pull/2/files#diff-dbaded28d0aac63950577de3ba365815d3f01c538ac3746e364fce47b643dcb4R23). Did you open an issue with PapaParse as I have suggested? (Although I don't believe there's a bug in the code, but that you don't fully understand the quoting and escaping Char options. But this is also something that you can probably find out with a ticket in the papaparse repo.)

Sure. Yes, i'm a little confused. I'll do that ASAP

**Comment by coderrsid** [comment]
> Please do not force-push. This makes reviewing and following the commits harder as necessary.

Actually i just removed last commit there. Won't do that anymore

**Comment by PackElend** [comment]
> i'm a little confused.

why? 
what confuses you?

**Comment by coderrsid** [comment]
> > i'm a little confused.
> 
> why?
> what confuses you?

The differentiation between escapeChar and quoteChar, but now i do. Anyways @tessus , i got a reply from them https://github.com/mholt/PapaParse/issues/885#event-4976532096 . They suggested the same method of removing characters like i'm using right now.

image

**Comment by tessus** [comment]
Ok, in this case, you can merge for now.

However, we need a lot more test cases and feedback from the community.
e.g. there are still no test cases with empty lines in the csv input.

**Review by tessus — APPROVED** [review]
We can merge, but we need more test cases.

Furthermore a plan on which options to use and why.

**Comment by tessus** [comment]
Btw, it would also be nice if you could setup a gh action that tests are run.

**Comment by coderrsid** [comment]
> Ok, in this case, you can merge for now.

Thanks Helmut.

> However, we need a lot more test cases and feedback from the community.
> e.g. there are still no test cases with empty lines in the csv input.

Yes, that is absolutely what week 4 is and will be (i.e. testing the base structure and the  full functionality of plugin)

**Comment by coderrsid** [comment]
> Furthermore a plan on which options to use and why.

Yes, we can discuss that on Discord.

**Comment by coderrsid** [comment]
> Btw, it would also be nice if you could setup a gh action that tests are run.

I'll add that effectively.
